packages/go_router/lib/src/route.dart

```diff
@@ -227,3 +145,55 @@ class GoRoute {
   ) =>
       const SizedBox.shrink();
 }
+
+/// A route that displays a UI shell around the matching child route.
+///
+/// The widget built by the matching child route becomes to the child parameter
+/// of the [builder].
+class ShellRoute extends RouteBase {
+  /// Constructs a [ShellRoute].
+  ShellRoute({
+    required String path,
+    required this.builder,
+    this.defaultRoute,
+    GoRouterRedirect redirect = RouteBase._emptyRedirect,
+    List<RouteBase> routes = const <RouteBase>[],
+  }) : super._(
+          path: path,
+          routes: routes,
+          redirect: redirect,
+        );
+
+  /// The widget builder for a shell route.
+  final ShellRouteBuilder builder;
+
+  /// The relative path to the child route to navigate to when this route is
+  /// displayed. This allows the default child route to be specified without
+  /// using redirection.
+  final String? defaultRoute;
+}
+
+/// A route that displays all descendent [StackedRoute]s within its visual
+/// boundary, typically the UI shell of a [ShellRoute].
+///
+/// This route places a nested [Navigator] in the widget tree, where any
+/// descendent [StackedRoute]s are placed onto this
+/// Navigator instead of the root Navigator, which allows you to display a UI
+/// shell around a nested stack of routes if this route is a child route of
+/// [ShellRoute].
+class NestedStackRoute extends RouteBase {
+  /// Constructs a [NestedRoute].
+  NestedStackRoute({
+    required String path,
+    required this.builder,
+    GoRouterRedirect redirect = RouteBase._emptyRedirect,
+    List<RouteBase> routes = const <RouteBase>[],
+  }) : super._(
+          path: path,
+          routes: routes,
+          redirect: redirect,
+        );
+
+  /// The widget builder for a nested stack route.
+  final NestedStackRouteBuilder builder;
```

Comment:
Maybe i missing something. should the function signature be the same as ShellRouteBuilder? Otherwise, how would someone builds a Shell-like UI

Reply:
They need to combine ShellRoute with NestedStackRoute see go_router_prototype for examples

Reply:
Since it's the same function type, I took this out.